File: vm/src/main/java/com/libreshockwave/vm/util/StringUtils.java
package com.libreshockwave.vm.util;

/**
 * Utility methods for string manipulation in display/formatting contexts.
 */
public final class StringUtils {

    private StringUtils() {}

    /**
     * Truncate a string if it exceeds the maximum length.
     * Appends "..." to indicate truncation.
     *
     * @param s the string to truncate (may be null)
     * @param maxLength maximum length including the "..." suffix
     * @return truncated string, or empty string if input is null
     */
    public static String truncate(String s, int maxLength) {
        if (s == null) return "";
        if (s.length() <= maxLength) return s;
        return s.substring(0, maxLength - 3) + "...";
    }

    /**
     * Escape special characters for display purposes.
     * Converts newlines, carriage returns, tabs, and backslashes to escape sequences.
     *
     * @param s the string to escape (may be null)
     * @return escaped string, or empty string if input is null
     */
    public static String escapeForDisplay(String s) {
        if (s == null) return "";
        return s.replace("\\", "\\\\")
                .replace("\n", "\\n")
                .replace("\r", "\\r")
                .replace("\t", "\\t");
    }

    /**
     * Escape special characters for HTML display.
     * Converts &amp;, &lt;, and &gt; to their HTML entity equivalents.
     *
     * @param s the string to escape (may be null)
     * @return HTML-escaped string, or empty string if input is null
     */
    public static String escapeHtml(String s) {
        if (s == null) return "";
        return s.replace("&", "&amp;")
                .replace("<", "&lt;")
                .replace(">", "&gt;");
    }
}

File: cpp/src/util/StringUtils.cpp
#include "libreshockwave/util/StringUtils.hpp"

namespace libreshockwave::util {

std::string truncate(std::string_view value, int maxLength) {
    if (maxLength >= 0 && value.size() <= static_cast<std::size_t>(maxLength)) {
        return std::string(value);
    }
    if (maxLength <= 3) {
        return "...";
    }
    return std::string(value.substr(0, static_cast<std::size_t>(maxLength - 3))) + "...";
}

std::string escapeForDisplay(std::string_view value) {
    std::string escaped;
    escaped.reserve(value.size());
    for (char ch : value) {
        switch (ch) {
            case '\\':
                escaped += "\\\\";
                break;
            case '\n':
                escaped += "\\n";
                break;
            case '\r':
                escaped += "\\r";
                break;
            case '\t':
                escaped += "\\t";
                break;
            default:
                escaped.push_back(ch);
                break;
        }
    }
    return escaped;
}

std::string escapeHtml(std::string_view value) {
    std::string escaped;
    escaped.reserve(value.size());
    for (char ch : value) {
        switch (ch) {
            case '&':
                escaped += "&amp;";
                break;
            case '<':
                escaped += "&lt;";
                break;
            case '>':
                escaped += "&gt;";
                break;
            default:
                escaped.push_back(ch);
                break;
        }
    }
    return escaped;
}

} // namespace libreshockwave::util
